lib/utils.js

```diff
@@ -1,246 +1,212 @@
+ * to use for GitHub API requests.
  *
- * @param  {Object} `options`
- * @return {Object}
+ * @param  {Object} `appOptions` Options on the instance of github-base.
+ * @return {Function} Returns a function that takes a request method name, the URL to expand, and method options.
  */
 
-utils.defaults = function(options) {
-  return function(method, url, data) {
-    var opts = utils.extend({method: 'get'}, options);
-    opts.apiurl = opts.apiurl || 'https://api.github.com';
-    var config = lowercaseKeys(opts);
-    opts = utils.merge({}, config, data);
-    opts = interpolate(url, opts);
-    opts = createURL(opts);
-    opts = lowercaseKeys(opts);
-    opts = createAuth(opts);
-    opts = cleanup(opts);
-    opts.method = method.toLowerCase();
-    opts = normalizeOpts(opts, config);
-
-    var defaults = {
-      'accept': 'application/json',
-      'user-agent': 'https://github.com/jonschlinkert/github-base'
-    };
-
-    opts.headers = utils.extend({}, defaults, lowercaseKeys(opts.headers));
-    return opts;
-  };
-};
+exports.defaults = function(appOptions) {
+  const defaults = { apiurl: 'https://api.github.com', method: 'get', params: [] };
 
-/**
- * Replace params with actual values.
- */
+  return function(method, uri, methodOptions) {
+    const options = Object.assign({}, defaults, appOptions, methodOptions, { uri, method });
+    const context = { orig: clone(options), options: lowercaseKeys(options) };
+    context.options.url = interpolate(uri, context);
+    context.options.headers = createHeaders(context.options);
 
-function interpolate(path, options) {
-  var opts = utils.extend({}, options, {params: []});
-  opts.url = path.replace(/:([\w_]+)/g, function(m, prop) {
-    opts.params.push(prop);
-    return opts[prop] || prop;
-  });
-  return opts;
-}
+    let opts = context.options;
+    const body = omit(context.orig, Object.keys(opts));
+    const json = opts.json;
+    const text = opts.body;
 
-/**
- * Create url to request and prevent cache
- */
+    if (!opts.mode && /\/markdown(?!\/raw)/.test(opts.url)) {
+      opts.mode = 'gfm';
+    }
 
-function createURL(opts) {
-  opts.url += setPrefix(opts.url);
-  opts.url += (new Date()).getTime();
-  opts.url = opts.apiurl + opts.url;
-  return opts;
-}
+    const bodyKeys = Object.keys(body);
+    if (bodyKeys.length > 0) {
+      opts = omit(opts, bodyKeys);
+      opts.headers['content-length'] = JSON.stringify(body).length;
 
-/**
- * Lowercase the keys in the given object.
- */
+      if (typeof opts.headers['content-type'] !== 'string') {
+        opts.headers['content-type'] = 'application/json';
+      }
+    } else if (lowercase(opts.method) === 'put') {
+      opts.headers['content-length'] = 0;
+    }
 
-function lowercaseKeys(obj) {
-  var res = {};
-  for (var key in obj) res[key.toLowerCase()] = obj[key];
-  return res;
-}
+    if (text) {
+      opts.body = text;
+    } else if (opts.body && opts.json !== true) {
```

Comment:
as mentioned above, since `text` and `opts.body` aren't modified before getting here, if `opts.body` has a value originally, then this `else if` will never happen.

Reply:
see my previous comment.